esiclient/v1/orchestrator/openshift.py

```diff
@@ -0,0 +1,339 @@
+#   a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#   WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#   License for the specific language governing permissions and limitations
+#   under the License.
+
+import json
+import logging
+import os
+import requests
+import time
+
+from osc_lib.command import command
+from osc_lib.i18n import _
+
+from esiclient import utils
+
+
+BASE_ASSISTED_INSTALLER_URL = \
+    'https://api.openshift.com/api/assisted-install/v2/'
+
+
+def call_assisted_installer_api(url, method, headers={}, data=None):
+    full_url = BASE_ASSISTED_INSTALLER_URL + url
+    if method == 'post':
+        response = requests.post(full_url, headers=headers, json=data)
+    elif method == 'get':
+        response = requests.get(full_url, headers=headers)
+    elif method == 'patch':
+        response = requests.patch(full_url, headers=headers, json=data)
+    else:
+        raise OSAIException(
+            'Unknown method used when calling Assisted Installer')
+    if response.status_code not in [200, 201, 202, 204]:
+        raise OSAIException(
+            "Unexpected response from Assisted Installer (%s): %s"
+            % (response.status_code, response.reason))
+    return response
+
+
+def wait_for_nodes(infra_env_id, headers, num_nodes, target_status):
+    waiting = True
+    print("waiting for hosts to reach %s..." % target_status)
+    while waiting:
+        response = call_assisted_installer_api(
+            "infra-envs/%s/hosts" % infra_env_id,
+            'get', headers)
+        count = 0
+        print("* checking hosts")
+        for host in response.json():
+            status = host.get('status')
+            print("  * " + host.get('requested_hostname') + ": " + status)
+            if status == target_status:
+                count = count + 1
+        if count == num_nodes:
+            waiting = False
+        else:
+            time.sleep(30)
+    print("... hosts ready")
+    return
+
+
+class OSAIException(Exception):
+    pass
+
+
+class Orchestrate(command.Lister):
+    """Orchestrate an OpenShift cluster using the Assisted Installer API"""
+
+    log = logging.getLogger(__name__ + ".Orchestrate")
+
+    def _print_failure_message(self, cluster_config_file, cluster_id=None,
+                               infra_env_id=None, message=None):
+        flags = ""
+        if cluster_id:
+            flags = "--cluster-id %s" % cluster_id
+            if infra_env_id:
+                flags = "%s --infra-env-id %s" % (flags, infra_env_id)
+        command = "openstack esi orchestrate openshift %s %s" % \
+            (flags, cluster_config_file)
+        if message:
+            print(message)
+        print("You may need to refresh your OpenShift token.")
```

Comment:
I think it might make users confused if they get an error that is not caused by token expiration, but still get the token refresh hint message. Then they just ignore the real error and rerun 'esi orchestrate openshift' anyway.

Reply:
Good point - fixed!